File: CPP/algorithms/backtracking/sudokuSolver.cpp
/*
Algorithm Name
Sudoku Solver

Programming Language
C++

Category
Other

Difficulty Level
Medium (Intermediate)

Algorithm Description
The Sudoku Solver problem requires filling a partially completed 9×9 Sudoku board so that each row, column, and 3×3 sub-grid contains digits 1–9 exactly once. The solution uses a recursive backtracking approach: we pick the next empty cell, try placing digits 1–9, and check validity against Sudoku rules. If a number fits, we recurse to solve the rest of the board; if it leads to a conflict later, we backtrack by removing it and trying the next digit. This process continues until the board is completely filled with a valid configuration.
*/
#include <bits/stdc++.h>
using namespace std;

bool isValid(vector<vector<char>>& board, int row, int col, char num) {
    // Check row and column
    for (int i = 0; i < 9; i++) {
        if (board[row][i] == num) return false;
        if (board[i][col] == num) return false;
    }

    // Check 3x3 sub-grid
    int startRow = 3 * (row / 3);
    int startCol = 3 * (col / 3);
    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            if (board[startRow + i][startCol + j] == num) return false;

    return true;
}

bool solveSudoku(vector<vector<char>>& board) {
    for (int row = 0; row < 9; row++) {
        for (int col = 0; col < 9; col++) {
            if (board[row][col] == '.') { // empty cell
                for (char num = '1'; num <= '9'; num++) {
                    if (isValid(board, row, col, num)) {
                        board[row][col] = num;
                        if (solveSudoku(board))
                            return true;
                        board[row][col] = '.'; // backtrack
                    }
                }
                return false; // no valid number found
            }
        }
    }
    return true; // solved
}

void printBoard(const vector<vector<char>>& board) {
    for (auto& row : board) {
        for (auto c : row)
            cout << c << " ";
        cout << endl;
    }
}

int main() {
    vector<vector<char>> board = {
        {'5','3','.','.','7','.','.','.','.'},
        {'6','.','.','1','9','5','.','.','.'},
        {'.','9','8','.','.','.','.','6','.'},
        {'8','.','.','.','6','.','.','.','3'},
        {'4','.','.','8','.','3','.','.','1'},
        {'7','.','.','.','2','.','.','.','6'},
        {'.','6','.','.','.','.','2','8','.'},
        {'.','.','.','4','1','9','.','.','5'},
        {'.','.','.','.','8','.','.','7','9'}
    };

    if (solveSudoku(board))
        printBoard(board);
    else
        cout << "No solution exists" << endl;

    return 0;
}

/*
Time Complexity: 
O(9^m) where m = empty cells.

Space Complexity: 
O(m) recursion stack.
*/


/*
Output
5 3 4 6 7 8 9 1 2 
6 7 2 1 9 5 3 4 8 
1 9 8 3 4 2 5 6 7 
8 5 9 7 6 1 4 2 3 
4 2 6 8 5 3 7 9 1 
7 1 3 9 2 4 8 5 6 
9 6 1 5 3 7 2 8 4 
2 8 7 4 1 9 6 3 5 
3 4 5 2 8 6 1 7 9 
*/

File: Python/algorithms/backtracking/sudoku_solver.py
"""
Algorithm: Sudoku Solver
Description: Solves a 9x9 Sudoku puzzle using backtracking
Time Complexity: O(9^(n*n)) where n=9, exponential in worst case
Space Complexity: O(n*n) for recursion stack
Author: DSA Code Repository
Date: 2025-10-02
"""

def is_valid(board, row, col, num):
    """
    Check if placing num at (row, col) is valid in Sudoku rules.
    
    Args:
        board: 9x9 Sudoku grid
        row: Row index (0-8)
        col: Column index (0-8)
        num: Number to place (1-9)
        
    Returns:
        bool: True if placement is valid, False otherwise
    """
    # Check row
    for x in range(9):
        if board[row][x] == num:
            return False
    
    # Check column
    for x in range(9):
        if board[x][col] == num:
            return False
    
    # Check 3x3 box
    start_row = row - row % 3
    start_col = col - col % 3
    for i in range(3):
        for j in range(3):
            if board[i + start_row][j + start_col] == num:
                return False
    
    return True

def solve_sudoku(board):
    """
    Solve Sudoku puzzle using backtracking.
    
    Args:
        board: 9x9 list of lists, 0 represents empty cells
        
    Returns:
        bool: True if solution exists, False otherwise
        
    Examples:
        >>> board = [[5,3,0,0,7,0,0,0,0],
        ...          [6,0,0,1,9,5,0,0,0],
        ...          [0,9,8,0,0,0,0,6,0],
        ...          [8,0,0,0,6,0,0,0,3],
        ...          [4,0,0,8,0,3,0,0,1],
        ...          [7,0,0,0,2,0,0,0,6],
        ...          [0,6,0,0,0,0,2,8,0],
        ...          [0,0,0,4,1,9,0,0,5],
        ...          [0,0,0,0,8,0,0,7,9]]
        >>> solve_sudoku(board)
        True
    """
    # Find empty cell
    for i in range(9):
        for j in range(9):
            if board[i][j] == 0:
                # Try numbers 1-9
                for num in range(1, 10):
                    if is_valid(board, i, j, num):
                        # Make choice
                        board[i][j] = num
                        
                        # Recursive call
                        if solve_sudoku(board):
                            return True
                        
                        # Backtrack
                        board[i][j] = 0
                
                return False
    
    return True  # All cells filled

def print_board(board):
    """Print Sudoku board in a formatted way."""
    for i in range(9):
        if i % 3 == 0 and i != 0:
            print("- - - - - - - - - - - -")
        
        for j in range(9):
            if j % 3 == 0 and j != 0:
                print(" | ", end="")
            
            if j == 8:
                print(board[i][j])
            else:
                print(str(board[i][j]) + " ", end="")

def main():
    """Test the Sudoku solver with example puzzles."""
    print("Sudoku Solver using Backtracking")
    print("=" * 40)
    
    # Test case 1: Easy puzzle
    board = [
        [5, 3, 0, 0, 7, 0, 0, 0, 0],
        [6, 0, 0, 1, 9, 5, 0, 0, 0],
        [0, 9, 8, 0, 0, 0, 0, 6, 0],
        [8, 0, 0, 0, 6, 0, 0, 0, 3],
        [4, 0, 0, 8, 0, 3, 0, 0, 1],
        [7, 0, 0, 0, 2, 0, 0, 0, 6],
        [0, 6, 0, 0, 0, 0, 2, 8, 0],
        [0, 0, 0, 4, 1, 9, 0, 0, 5],
        [0, 0, 0, 0, 8, 0, 0, 7, 9]
    ]
    
    print("Original Sudoku puzzle:")
    print_board(board)
    print()
    
    if solve_sudoku(board):
        print("Solved Sudoku puzzle:")
        print_board(board)
    else:
        print("No solution exists for this Sudoku puzzle!")
    
    print("\n" + "=" * 40)
    
    # Test case 2: Empty board
    empty_board = [[0 for _ in range(9)] for _ in range(9)]
    print("Solving empty Sudoku board...")
    
    if solve_sudoku(empty_board):
        print("Solution for empty board:")
        print_board(empty_board)
    else:
        print("No solution found!")

if __name__ == "__main__":
    main()